Initial set of commands

stbenjam opened the pull request:
This is the first round of commands to handle all the basics of working with remote execution.  Some additional API's and other changes are needed, those are provided by https://github.com/theforeman/foreman_remote_execution/pull/78

See --help and also the README here for some examples.

To do:
- [x] Tests (Discovery has some good examples, https://github.com/theforeman/hammer-cli-foreman-discovery/tree/master/test/unit)

Template commands:
- [x] Info
- [x] Dump
- [x] List
- [x] Create
- [x] Update
- [x] Delete
- [x] Template Input subcommand (needs new API)
  - [x] Info
  - [x] List
  - [x] Create
  - [x] Delete

Invocation commands:
- [x] List
- [x] Create
- [x] Info
- [x] Output (command output - live if the job is still running)

stbenjam commented:
Obligatory gif

![rex](https://cloud.githubusercontent.com/assets/429763/11367480/e2e84328-9281-11e5-9825-364cfd4765cd.gif)

Still a few rough areas, but it's working pretty well now

iNecas commented:
I'm hitting this issue right now http://pastebin.com/0Aefx3nK, how have you solved it for you?

iNecas commented:
Oh… updating the hammer-cli-formen seemed to fix that :)

iNecas commented:
Very nice! Some usability comments:

* the `targeting_type` would be better with some defaults and not required (especially for immediate execution, the dynamic targeting has just a little sense). perhaps static targeting by default
* it would be nice to see a list of hosts that were targeted for the execution, ideally showing that list (perhaps with adjustable pagination option) - the desired workflow - make it easy to trigger get the output output based on the output of the invocation command, maybe with printing a message like:
```
run `hammer job-invocation output --id 144 host host-1.foreman.test` to see the output of an output
```
* I wonder, if `hammer job-invocation output --async` would make sense, to get the current output, but don't block
* ability to filter the hosts for the job invocation (also, filtering the failed hosts would be interesting, but we don't have that in UI anyway, so more work would be needed to be done there)
* re-run based on previous invocation (rerun failed)

iNecas commented on README.md:
````diff
@@ -1,8 +1,42 @@
+
+```
+hammer job-invocation create --job-name "Ping" --job-template "Ping a Host"\
+  --inputs hostname=www.google.com --search-query "name ~ rex01"\
````
An example on how to properly pass multiple inputs would be nice

adamruzicka replied:
And also how to properly escape quotation marks would come in handy since naive `--input "command=echo 'this is a test'"` somehow makes the second `'` disappear

iNecas commented on lib/hammer_cli_foreman_remote_execution/job_template.rb:
```diff
@@ -0,0 +1,101 @@
+
+      if template['template_inputs']
+        new_input = "\n"
+        template['template_inputs'].each { |option| new_input += "  - #{option['name']}\n" }
```
I would prefer
```ruby
template_inputs = template['template_inputs'].map { |option| "  - #{option['name']}" }.join("\n")
template['template_inputs'] = "\n#{template_inputs}\n"
```

stbenjam commented:
> - it would be nice to see a list of hosts that were targeted for the execution, ideally showing that list (perhaps with adjustable pagination option) - the desired workflow - make it easy to trigger get the output output based on the output of the invocation command

> - ability to filter the hosts for the job invocation (also, filtering the failed hosts would be interesting, but we don't have that in UI anyway, so more work would be needed to be done there)

These two comments are somewhat related. I want to show more info in `job-invocation info`, like an ascii table that includes the list of hosts and a summary of their respective status.  But we need to do something better about how we store the data, because right now it's very difficult to search or filter on it.

We use a lot of view helpers to dig the data out for the UI, but there has to be a better way that's queriable, sortable, filterable, etc.

We need it to handle http://projects.theforeman.org/issues/12571, too.  It's pretty frustrating for end users that very few of our columns are sortable/searchable because its buried deep in the foreman tasks subtask.  I'm not sure what the best way to do it is - could we bubble the data up as columns on the join model (TargetingHost)? That would make scoped search Just Work a lot of places.  Has Katello ever solved this anywhere maybe?

stbenjam commented:
Other than that, thanks! Great feedback!

stbenjam commented:
Forcing travis to test

stbenjam commented:
Ok, done here:

- Dynamic is now an optional flag (`--dynamic`) instead of this `--targeting-query=dynamic_query` thingy
- `--async` option for job invocation output
- Added examples for multiple inputs (`--inputs package=vim-enhanced,action=install`) to the README

- @adamruzicka had a good point about complex inputs, it's hard to get right on the command line.  I've added a way to read inputs from files.  It works like this:

```
hammer job-invocation create --job-name "Run Command"\
  --input-files command=/tmp/script.sh --search-query "name ~ rex01"
```

In this case command would be read from `/tmp/script.sh`.  A more complex example using both:

```
hammer job-invocation create --job-name "Run Command"\
  --inputs package=vim-enhanced,action=install
  --input-files pre_script=pre_script.sh,post_script=post_script.sh --search-query "name ~ rex01"
```

The other thought I had was a `--input-yaml` option, and put all the inputs in a single YAML file.  IMHO, sysadmins probably prefer just being able to pass a single text file for an input.

stbenjam commented:
Tests with VCR are kind of unstable. I'll have another look at the Apipie resource mocks, but I couldn't really figure out how to get them working.

adamruzicka commented:
The `--input-files command=./somefile` works nicely, being able to pass in a single YAML file overridable by from the cli would be nice.

Would it be possible to have the output command grab outputs from all hosts and do something like `tail(1)` does when the user doesn't provide host name/id? If not, it would be useful to have some kind of host listing in the `info` command.

iNecas commented:
the tail over all hosts might produce quite a lot of trafic between foreman and proxy -  we would need to optimize the communication there first, I'm ok with not doing so on first-pass

iNecas commented:
Small usably suggestion: the provider-type could be case-insensitive, and maybe default to Ssh when not specified

iNecas commented:
So for the template inputs, the plan right now is just use the ids for identifying the templates are addressing that after merging the this PR or wait for it for? Ability to add the input without needing to know the template id would be a great improvament

iNecas commented on lib/hammer_cli_foreman_remote_execution.rb:
```diff
@@ -1,4 +1,10 @@
+require 'hammer_cli_foreman_tasks'
 
 module HammerCLIForemanRemoteExecution
+  require 'hammer_cli_foreman_remote_execution/options/normalizers'
```
I've noticed the hammer was not displaying error messages from the validations. Adding this helped the situation.

```
  def self.exception_handler_class
    HammerCLIForeman::ExceptionHandler
  end
```

stbenjam replied:
Ah thanks! Much better

stbenjam commented:
> So for the template inputs, the plan right now is just use the ids for identifying the templates are addressing that after merging the this PR or wait for it for? Ability to add the input without needing to know the template id would be a great improvament

I agree, but the problem is the names aren't even unique at the Template level. It's probably likely they are but we don't make that constraint.  We'd also need a Template api to handle the searches. And figure out how to deal with the fact that hammer-cli-foreman thinks 'template' means only 'provisioning template'.  It's not a good situation. :disappointed: 

I could hack it so template-input only looks at job-templates, but if anyone ever uses it for something else, it won't work.  Who did TemplateInput, @ares? 

@ares, Do you have a vision for how this should work? :smile:

iNecas commented on lib/hammer_cli_foreman_remote_execution/job_invocation.rb:
```diff
@@ -0,0 +1,112 @@
+    end
+
+    class OutputCommand < HammerCLIForeman::Command
+      action :output
```
The output action has the following option, that I'm not sure what to use for:

```
 --name NAME                   Name to search by
```

stbenjam replied:
The hammer docs theoretically say I can prevent this with `build_options { |o| o.expect(:name) }` but it doesn't work.  On IRC, the hammer people (I forget who), told me I probably need to write an entire Id resolver implementation like hammer-cli-katello, which is very unfortunate.

iNecas replied:
----- Original Message -----
> > +    class InfoCommand < HammerCLIForeman::InfoCommand
> > +      output ListCommand.output_definition do
> > +        field :hosts, _('Hosts')
> > +      end
> > +
> > +      def extend_data(invocation)
> > +        JobInvocation.extend_data(invocation)
> > +      end
> > +
> > +      build_options do |o|
> > +        o.expand(:none)
> > +      end
> > +    end
> > +
> > +    class OutputCommand < HammerCLIForeman::Command
> > +      action :output
> 
> The hammer docs theoretically say I can prevent this with `build_options {
> |o| o.expect(:name) }` but it doesn't work.  On IRC, the hammer people (I
> forget who), told me I probably need to write an entire Id resolver
> implementation like hammer-cli-katello, which is very unfortunate.

I would rather track this as a bug and fixing in hammer rather than us duplicating the
code.

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/theforeman/hammer_cli_foreman_remote_execution/pull/1/files#r47103498

stbenjam replied:
http://projects.theforeman.org/issues/12767

iNecas commented on lib/hammer_cli_foreman_remote_execution/job_invocation.rb:
```diff
@@ -0,0 +1,112 @@
+    class CreateCommand < HammerCLIForeman::CreateCommand
+      include HammerCLIForemanTasks::Async
+
+      success_message _('Job invocation %{id} created')
```
`_('Job invocation %{id} started')` would be more appropriate

iNecas commented on lib/hammer_cli_foreman_remote_execution/job_invocation.rb:
```diff
@@ -0,0 +1,112 @@
+      end
+
+      def task_progress(task_or_id)
+        task = task_or_id['dynflow_task']['id']
```
adding `print_message(success_message, task_or_id)` would print the job invocation id - this would be pretty handy once you need the id of the job invocation (info, output).

iNecas commented:
I'm ok with merging this and addressing the dependent things in further PRs.

iNecas commented:
I've tested it and it works great

stbenjam commented:
@iNecas Squashed and green

iNecas commented:
ACK
